src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -643,6 +644,16 @@ private void AddEnvDTEProjectToCache(EnvDTE.Project envDTEProject)
             // Then create the project name from the project.
             var newEnvDTEProjectName = ProjectNames.FromDTEProject(envDTEProject);
 
+            PackageSpec packageSpec;
+
+            _projectSystemCache.TryGetProjectRestoreInfo(newEnvDTEProjectName.FullName, out packageSpec);
```

Comment:
Move the whole thing (lines 647-656) to the `ProjectSystemCache.AddProject`. It should be transactional and not visible to callers.